### PR TITLE
Adding service-queue to mgmtworker and add/remove manager to/from clu…

### PR DESCRIPTION
**Pull request by barucoh**
…ster handlers

### CODE REVIEW

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    from cloudify_premium.ha import syncthing
+except ImportError:
+    update_devices = None
+
+logger = logging.getLogger('worker.mgmtworker')
+
+
+class ClusterServiceConsumer(TaskConsumer):
+    """
+    This consumer is intended to be used only by the mgmt-worker and only for
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: 'mgmt-worker' we normally just write in one work mgmtworker

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        task_name = task['task_name']
+        kwargs = task['kwargs']
+
+        logger.info(
+            'Received `{0}` cluster service task'.format(
+                task_name
+            )
+        )
+        result = self.tasks_map[task_name](**kwargs)
+        logger.info('Result: {0}'.format(result))
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: maybe add the task name also here to verify the context of the log msg

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    def manager_added(self):
+        logger.info('A manager has been added to the cluster')
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: maybe add a log msg to the end also and fix this msg accordingly

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        syncthing.update_devices()
+
+    def manager_removed(self):
+        logger.info('A manager has been removed from the cluster')
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: maybe add a log msg to the end also and fix this msg accordingly

File: rest-service/manager_rest/rest/resources_v3_1/manager.py [hunk 5]
```diff
@@ -142,7 +149,10 @@ def post(self):
             distro_release=_manager['distro_release'],
             fs_sync_node_id=_manager.get('fs_sync_node_id', '')
         )
-        return get_storage_manager().put(new_manager)
+        result = get_storage_manager().put(new_manager)
+        if result['hostname'] == result['hostname']:  # Object created
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: use new manager

File: rest-service/manager_rest/rest/resources_v3_1/manager.py [hunk 6]
```diff
@@ -181,8 +191,10 @@ def delete(self):
             filters={'hostname': _manager['hostname']}
         )
 
-        # TODO: send message on service-queue
-        return sm.delete(manager_to_delete)
+        result = sm.delete(manager_to_delete)
+        if result['hostname'] == result['hostname']:  # Object deleted
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: use manager param

Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: also the comment is unclear

File: rest-service/manager_rest/rest/resources_v3_1/manager.py [hunk 7]
```diff
@@ -181,8 +191,10 @@ def delete(self):
...
+            remove_manager()  # Removing manager from cluster
+        return result
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: maybe add logging before and after

File: rest-service/manager_rest/rest/resources_v3_1/manager.py [hunk 8]
```diff
@@ -142,7 +149,10 @@ def post(self):
             distro_release=_manager['distro_release'],
             fs_sync_node_id=_manager.get('fs_sync_node_id', '')
         )
-        return get_storage_manager().put(new_manager)
+        result = get_storage_manager().put(new_manager)
+        if result['hostname'] == result['hostname']:  # Object created
+            add_manager(result['hostname'])  # Adding manager to cluster
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: maybe add logging before and after

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import logging
+
+from cloudify.constants import CLUSTER_SERVICE_EXCHANGE_NAME
+from cloudify.amqp_client import TaskConsumer
+try:
+    from cloudify_premium.ha import syncthing
+except ImportError:
+    update_devices = None
+
+logger = logging.getLogger('worker.mgmtworker')
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: why worker? and not 'mgmtworker.cluster'?

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py (hunk 1 above)
Comment (barucoh) [reply mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: Fixed

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py (hunk 2 above)
Comment (barucoh) [reply mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: Fixed

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py (hunk 3 above)
Comment (barucoh) [reply mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: I updated the message and added another one on Syncthing side when the operation finished

File: rest-service/manager_rest/rest/resources_v3_1/manager.py (hunk 5 above)
Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3_1/manager.py]: Fixed

File: rest-service/manager_rest/rest/resources_v3_1/manager.py (hunk 6 above)
Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3_1/manager.py]: Fixed

Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3_1/manager.py]: Fixed

File: rest-service/manager_rest/rest/resources_v3_1/manager.py (hunk 8 above)
Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3_1/manager.py]: Fixed

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py (hunk 9 above)
Comment (barucoh) [reply mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: Fixed

File: rest-service/manager_rest/rest/resources_v3_1/manager.py (hunk 7 above)
Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3_1/manager.py]: Fixed

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+############
+
+import logging
+
+from cloudify.constants import CLUSTER_SERVICE_EXCHANGE_NAME
+from cloudify.amqp_client import TaskConsumer
+try:
+    from cloudify_premium.ha import syncthing
+except ImportError:
+    update_devices = None
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: Why is this needed? if there is no prem this will crash anyway

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        task_name = task['task_name']
+        kwargs = task['kwargs']
+
+        logger.info(
+            'Received `{0}` cluster service task'.format(
+                task_name
+            )
+        )
+        result = self.tasks_map[task_name](**kwargs)
+        logger.info('{0} Result: {1}'.format(task_name, result))
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: What about "{0}"?

File: mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    def manager_added(self):
+        logger.info('A manager has been added to the cluster, updating '
```
Comment (tyacbovi) [inline mgmtworker/mgmtworker/task_consumers/cluster_service_consumer.py]: what about the logs ordering?

File: rest-service/manager_rest/rest/resources_v3_1/manager.py [hunk 13]
```diff
@@ -142,7 +149,11 @@ def post(self):
             distro_release=_manager['distro_release'],
             fs_sync_node_id=_manager.get('fs_sync_node_id', '')
         )
-        return get_storage_manager().put(new_manager)
+        result = get_storage_manager().put(new_manager)
+        if _manager['hostname'] == result['hostname']:
+            current_app.logger.info('Manager added successfully')
```
Comment (tyacbovi) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: maybe switch places with the next line

File: mgmtworker/mgmtworker/worker.py [hunk 14]
```diff
@@ -42,14 +47,22 @@ def _setup_logger():
 def make_amqp_worker(args):
     operation_registry = ProcessRegistry()
     workflow_registry = ProcessRegistry()
+    operation_consumer = WaitingCloudifyOperationConsumer
+    workflow_consumer = WaitingCloudifyWorkflowConsumer
+    try:
```
Comment (tehasdf) [inline mgmtworker/mgmtworker/worker.py]: instead of this, i'd do inside of `wait_for_replication_mixin.py` something like

```python
try:
  import cloudify_premium
except ImportError:
  has_premium = False
else:
  has_premium = True
```
and check for that flag here
this way we'll only have one `try/except ImportError`, instead of two, and we'll lose the ugly import inside a func

File: mgmtworker/mgmtworker/task_consumers/wait_for_replication_mixin.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+)
+try:
+    from cloudify_premium import mgmtworker_utils
+except ImportError:
+    mgmtworker_utils = None
+
+logger = logging.getLogger('mgmtworker')
+
+
+class WaitForReplicationMixin(object):
```
Comment (tehasdf) [inline mgmtworker/mgmtworker/task_consumers/wait_for_replication_mixin.py]: not really needed now? or anyway, with the current premium-side impl

File: rest-service/manager_rest/rest/resources_v3/cluster.py [hunk 16]
```diff
@@ -20,9 +20,9 @@
 from ..responses_v3 import BaseResponse
 
 try:
-    from cloudify_premium import (ClusterResourceBase,
-                                  ClusterState,
-                                  ClusterNode)
+    from cloudify_premium.ha.web import (ClusterResourceBase,
```
Comment (tehasdf) [inline rest-service/manager_rest/rest/resources_v3/cluster.py]: eh maybe just remove the whole file

File: rest-service/manager_rest/rest/resources_v3_1/manager.py [hunk 17]
```diff
@@ -142,7 +149,11 @@ def post(self):
             distro_release=_manager['distro_release'],
             fs_sync_node_id=_manager.get('fs_sync_node_id', '')
         )
-        return get_storage_manager().put(new_manager)
+        result = get_storage_manager().put(new_manager)
+        current_app.logger.info('Manager added successfully')
+        if _manager.get('fs_sync_node_id'):
+            add_manager()
```
Comment (tehasdf) [inline rest-service/manager_rest/rest/resources_v3_1/manager.py]: you know, if we just sent the list of managers here, to the mgmtworker, then the mgmtworker-side task wouldn't need to query for it, and so wouldn't need the rest-client or the admin-token

File: rest-service/manager_rest/test/endpoints/test_managers.py [hunk 18]
```diff
@@ -21,27 +21,26 @@
...
-_manager2 = {
-    'hostname': 'manager2.test.domain',
-    'private_ip': '172.17.0.2',
-    'public_ip': '192.168.0.2',
-    'version': '5.0.dev1',
-    'edition': 'premium',
-    'distribution': 'centos',
-    'distro_release': 'Core'
-}
+_manager1 = dict(
```
Comment (tehasdf) [inline rest-service/manager_rest/test/endpoints/test_managers.py]: why this change? seems like just making the syntax worse for no gain :P

File: mgmtworker/mgmtworker/worker.py (hunk 14 above)
Comment (barucoh) [reply mgmtworker/mgmtworker/worker.py]: Fixed

File: mgmtworker/mgmtworker/task_consumers/wait_for_replication_mixin.py (hunk 15 above)
Comment (barucoh) [reply mgmtworker/mgmtworker/task_consumers/wait_for_replication_mixin.py]: I guess not, so I'll just remove the whole thing

File: rest-service/manager_rest/rest/resources_v3/cluster.py (hunk 16 above)
Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3/cluster.py]: I don't want to remove the old cluster together with this merge.
I have a separate task for it

File: rest-service/manager_rest/rest/resources_v3_1/manager.py (hunk 17 above)
Comment (barucoh) [reply rest-service/manager_rest/rest/resources_v3_1/manager.py]: I guess you're right.
Fixed

File: rest-service/manager_rest/test/endpoints/test_managers.py (hunk 18 above)
Comment (barucoh) [reply rest-service/manager_rest/test/endpoints/test_managers.py]: Yeah the tests failed at some point with something regarding 'dictionary' issues, so I thought this might change something.
I forgot to change it back :)
